librz/core/analysis_tp.c

```diff
@@ -266,6 +266,29 @@ static void retype_callee_arg(RzAnalysis *analysis, const char *callee_name, boo
 	}
 }
 
+static RzAnalysisOp *op_cache_seek(RzPVector *cache, ut64 addr, ssize_t *p_idx) {
+	ssize_t idx = *p_idx;
+	if (rz_pvector_empty(cache)) {
+		return NULL;
+	}
+	for (int i = 0; i < 10 && idx >= 0; i++, idx--) {
```

Comment:
isn't it faster to have a HT instead of a sorted Vector?

Reply:
No, we don't need random access. Actually this loop stops after 2 or even 1 iteration.

Reply:
why? You are looking for an entry in the cache with `op->addr == addr`, right? So you want to have immediate access to the element which has `addr` as key, don't you?

Reply:
As I understood `type_match()` analyzes instructions of a basic block in reverse order. The vector (aka cache) contains analysis data of all instructions of a BB. So we can just iterate over the vector in a reverse order. It is much faster than `ht_up_find()`, isn't it? And moreover we need to get two entries from the cache at every iteration. In case of a vector we can just use `vector[i]` + `vector[i+1]`. Why `op_cache_seek()` is needed? The vector can contain some entries which were not added to ESIL trace. So `op_cache_seek()` mostly just returns `vector[cache_idx]` but it can also decrease `cache_idx` value to skip some entries.